File: unique_api/app/router/metadata.py
import base64

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import rsa
from fastapi import APIRouter
from fastapi.responses import JSONResponse

from ..config import settings
from ..services.client_auth import ClientAuthMethod

router = APIRouter()


@router.get("/.well-known/openid-configuration")
async def openid_configuration():
    """
    OpenID Connect Discovery 1.0のメタデータエンドポイント
    https://openid.net/specs/openid-connect-discovery-1_0.html
    """
    metadata = {
        "issuer": settings.ISSUER,
        "authorization_endpoint": f"{settings.FRONTEND_URL}/auth",
        "token_endpoint": f"{settings.ISSUER}/token",
        "jwks_uri": f"{settings.ISSUER}/.well-known/jwks.json",
        "userinfo_endpoint": f"{settings.ISSUER}/userinfo",
        "scopes_supported": ["openid", "profile", "email"],
        "response_types_supported": ["code"],
        "grant_types_supported": ["authorization_code", "refresh_token"],
        "subject_types_supported": ["public"],
        "id_token_signing_alg_values_supported": [settings.JWT_ALGORITHM],
        "token_endpoint_auth_methods_supported": [
            ClientAuthMethod.CLIENT_SECRET_BASIC,
            ClientAuthMethod.CLIENT_SECRET_POST,
        ],
        "claims_supported": [
            "email",
            "email_verified",
            "name",
            "preferred_username",
            "iss",
            "sub",
            "aud",
            "exp",
            "iat",
            "auth_time",
            "nonce",
            "acr",
            "amr",
            "azp",
            "at_hash",
        ],
        "code_challenge_methods_supported": ["plain", "S256"],
    }

    return JSONResponse(
        content=metadata,
        headers={
            "Cache-Control": "public, max-age=3600",
            "Content-Type": "application/json",
        },
    )


@router.get("/.well-known/jwks.json")
async def jwks():
    """
    JSON Web Key Set (JWKS) エンドポイント
    https://openid.net/specs/openid-connect-discovery-1_0.html#JWKSEndpoint
    """
    jwks_data = {"keys": _generate_jwks_from_public_key()}  # 公開鍵からJWKSを生成

    return JSONResponse(
        content=jwks_data,
        headers={
            "Cache-Control": "public, max-age=3600",
            "Content-Type": "application/json",
        },
    )


def _generate_jwks_from_public_key() -> list:
    """
    公開鍵ファイルからJWKSを生成する
    """
    if not settings.RSA_PUBLIC_KEY_PATH:
        return []

    try:
        # 公開鍵を読み込む
        with open(settings.RSA_PUBLIC_KEY_PATH, "rb") as f:
            public_key_data = f.read()

        # 公開鍵をロード
        public_key = serialization.load_pem_public_key(public_key_data)

        # RSA公開鍵の場合のみ処理
        if isinstance(public_key, rsa.RSAPublicKey):
            public_numbers = public_key.public_numbers()

            # nとeをBase64url形式に変換
            def int_to_base64url(num: int) -> str:
                num_bytes = num.to_bytes((num.bit_length() + 7) // 8, byteorder="big")
                return base64.urlsafe_b64encode(num_bytes).decode("utf-8").rstrip("=")

            n = int_to_base64url(public_numbers.n)
            e = int_to_base64url(public_numbers.e)

            # JWKを構築
            jwk = {
                "kty": "RSA",
                "use": "sig",
                "alg": settings.JWT_ALGORITHM,
                "n": n,
                "e": e,
            }

            # kid（Key ID）を生成（オプション）
            # 公開鍵の指紋をsha256ハッシュで生成
            import hashlib

            key_id = hashlib.sha256(public_key_data).hexdigest()[:16]
            jwk["kid"] = key_id

            return [jwk]
        else:
            # RSA以外の鍵タイプには対応していない
            return []

    except (FileNotFoundError, ValueError, IOError) as e:
        # エラーログを出力するべきだが、ここでは空のリストを返す
        print(f"Error generating JWKS: {e}")
        return []

File: src/internal/router/metadata.go
package router

import (
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"

	"github.com/UniPro-tech/UniQUE-Auth/internal/config"
	"github.com/gin-gonic/gin"
)

type WellKnownOpenIDConfigurationResponse struct {
	Issuer                            string   `json:"issuer"`
	AuthorizationEndpoint             string   `json:"authorization_endpoint"`
	TokenEndpoint                     string   `json:"token_endpoint"`
	UserinfoEndpoint                  string   `json:"userinfo_endpoint"`
	RevocationEndpoint                string   `json:"revocation_endpoint"`
	JwksURI                           string   `json:"jwks_uri"`
	ResponseTypesSupported            []string `json:"response_types_supported"`
	SubjectTypesSupported             []string `json:"subject_types_supported"`
	IDTokenSigningAlgValuesSupported  []string `json:"id_token_signing_alg_values_supported"`
	ScopesSupported                   []string `json:"scopes_supported"`
	TokenEndpointAuthMethodsSupported []string `json:"token_endpoint_auth_methods_supported"`
	ClaimsSupported                   []string `json:"claims_supported"`
	CodeChallengeMethodsSupported     []string `json:"code_challenge_methods_supported"`
	GrantTypesSupported               []string `json:"grant_types_supported"`
}

// WellKnownOpenIDConfiguration godoc
// @Summary OpenID Configuration
// @Description .well-known/openid-configuration エンドポイント
// @Produce json
// @Tags metadata
// @Success 200 {object} WellKnownOpenIDConfigurationResponse
// @Router /.well-known/openid-configuration [get]
func WellKnownOpenIDConfiguration(c *gin.Context) {
	environmentConfigs := c.MustGet("config").(*config.Config)

	c.JSON(200, WellKnownOpenIDConfigurationResponse{
		Issuer:                            environmentConfigs.IssuerURL,
		AuthorizationEndpoint:             environmentConfigs.IssuerURL + "/authorization",
		TokenEndpoint:                     environmentConfigs.IssuerURL + "/token",
		UserinfoEndpoint:                  environmentConfigs.IssuerURL + "/userinfo",
		RevocationEndpoint:                environmentConfigs.IssuerURL + "/revocation",
		JwksURI:                           environmentConfigs.IssuerURL + "/.well-known/jwks.json",
		ResponseTypesSupported:            []string{"code" /* TODO: Impliment "code token", etc... */},
		SubjectTypesSupported:             []string{"public"},
		IDTokenSigningAlgValuesSupported:  []string{"RS256"},
		ScopesSupported:                   config.Scopes.AllowedScopes,
		TokenEndpointAuthMethodsSupported: []string{"client_secret_basic", "client_secret_post"},
		ClaimsSupported: []string{
			"sub",
			"iss",
			"aud",
			"exp",
			"iat",
			"name",
			"preferred_username",
			"email",
			"email_verified",
			// TODO: Impliment acr, amr, etc...
		},
		CodeChallengeMethodsSupported: []string{"S256"},
		GrantTypesSupported:           []string{"authorization_code", "refresh_token"},
	})
}

type JWKS struct {
	Keys []JWKSKey `json:"keys"`
}

type JWKSKey struct {
	Kty string `json:"kty"`
	Use string `json:"use"`
	Alg string `json:"alg"`
	Kid string `json:"kid"`
	N   string `json:"n"`
	E   string `json:"e"`
}

// WellKnownJWKS godoc
// @Summary JSON Web Key Set
// @Description .well-known/jwks.json エンドポイント
// @Produce json
// @Tags metadata
// @Success 200 {object} JWKS
// @Router /.well-known/jwks.json [get]
func WellKnownJWKS(c *gin.Context) {
	// Get Keys Path from config
	environmentConfigs := c.MustGet("config").(*config.Config)

	jwks, err := generateJWKSFromConfig(environmentConfigs)
	if err != nil {
		c.JSON(500, gin.H{"error": "internal_server_error"})
		return
	}
	c.JSON(200, jwks)
}

func generateJWKSFromConfig(cfg *config.Config) (*JWKS, error) {
	var keys []JWKSKey

	// iterate over configured key pairs
	for _, kp := range cfg.KeyPairs {
		rsaPub := &kp.PublicKey

		// marshal to PKIX DER to compute kid
		der, err := x509.MarshalPKIXPublicKey(rsaPub)
		if err != nil {
			continue
		}
		thumb := sha256.Sum256(der)
		kid := hex.EncodeToString(thumb[:])

		n := base64.RawURLEncoding.EncodeToString(rsaPub.N.Bytes())

		eInt := rsaPub.E
		eBytes := []byte{}
		for x := eInt; x > 0; x >>= 8 {
			eBytes = append([]byte{byte(x & 0xff)}, eBytes...)
		}
		if len(eBytes) == 0 {
			eBytes = []byte{0}
		}
		e := base64.RawURLEncoding.EncodeToString(eBytes)

		jwk := JWKSKey{
			Kty: "RSA",
			Use: "sig",
			Alg: "RS256",
			Kid: kid,
			N:   n,
			E:   e,
		}
		keys = append(keys, jwk)
	}

	return &JWKS{Keys: keys}, nil
}
